src/main/java/software/amazon/cloudformation/AbstractWrapper.java

```diff
@@ -150,7 +150,15 @@ public AbstractWrapper(final CredentialsProvider providerCredentialsProvider,
         // sync.
         // Both are required parameters when LoggingConfig (optional) is provided when
         // 'RegisterType'.
-        if (providerCredentials != null) {
+        if (providerCredentials == null) {
+            // reset provider credentials back to null to avoid reusing stale credentials
+            if (this.providerCredentialsProvider != null) {
+                this.providerCredentialsProvider.resetCredentials();
+            }
+            this.providerMetricsPublisher = null;
+            this.providerEventsLogger = null;
+            this.cloudWatchLogHelper = null;
+        } else {
```

Comment:
Why null check providerCredentials and providerCredentialsProvider before resetting the credentials? Cant we do it for every request? The next section will set it if credentials are provided anyway.

Reply:
because it does get refreshed when creds are non-null

Reply:
Actually that null check on providerCredentialsProvider seemed unnecessary to me.  It's a final variable and instantiated in the constructor, but it was already there and I was just afraid to remove it and potentially cause an NPE for something I couldn't see.

Reply:
Ok. What exactly are providerCredentials and providerCredentialsProvider? Do the credentials not need to be reset if both providerCredentials and providerCredentialsProvider are null?

Reply:
providerCredentials come directly from the request sent in by CloudFormation (or Cloud Control API).  So if the credentials passed by CFN are null, then we want to reset everything.  

ProviderCredentialsProvider is just a wrapper class which holds a reference to a credentials object.  Like I said I don't think it's possible to be null, but if it was then it wouldn't hold any credentials anyway, which is what we want in this case. And even then, we still reset the other related objects metricsPublisher, eventsLogger, and cloudWatchLogHelper